crypto/evp/s_lib.c

```diff
@@ -0,0 +1,310 @@
+    if (skey->keydata == NULL)
+        goto err;
+
+    return skey;
+
+ err:
+    EVP_SKEYMGMT_free(skeymgmt);
+    EVP_SKEY_free(skey);
+    return NULL;
+}
+
+EVP_SKEY *EVP_SKEY_generate(OSSL_LIB_CTX *libctx, const char *skeymgmtname,
+                            const char *propquery, const OSSL_PARAM *params)
+{
+    EVP_SKEYMGMT *skeymgmt = NULL;
+    EVP_SKEY *skey = evp_skey_alloc();
+
+    if (skey == NULL)
+        return NULL;
+
+    skeymgmt = EVP_SKEYMGMT_fetch(libctx, skeymgmtname, propquery);
+    if (skeymgmt == NULL) {
+        /*
+         * if the specific key_type is unkown, attempt to use the generic
+         * key management
+         */
+        skeymgmt = EVP_SKEYMGMT_fetch(libctx, OSSL_SKEY_TYPE_GENERIC, propquery);
+        if (skeymgmt == NULL) {
+            ERR_raise(ERR_LIB_EVP, ERR_R_FETCH_FAILED);
+            goto err;
+        }
+    }
+    skey->skeymgmt = skeymgmt;
+
+    skey->keydata = evp_skeymgmt_generate(skey->skeymgmt, params);
+    if (skey->keydata == NULL)
+        goto err;
+
+    return skey;
+
+ err:
+    EVP_SKEYMGMT_free(skeymgmt);
+    EVP_SKEY_free(skey);
+    return NULL;
+}
+
+struct raw_key_details_st {
+    const void **key;
+    size_t *len;
+};
+
+static int get_secret_key(const OSSL_PARAM params[], void *arg)
+{
+    const OSSL_PARAM *p = NULL;
+    struct raw_key_details_st *raw_key = arg;
+
+    if ((p = OSSL_PARAM_locate_const(params, OSSL_SKEY_PARAM_RAW_BYTES)) != NULL)
+        return OSSL_PARAM_get_octet_string_ptr(p, raw_key->key, raw_key->len);
+
+    return 0;
+}
+
+int EVP_SKEY_get_raw_key(const EVP_SKEY *skey, const unsigned char **key,
+                         size_t *len)
+{
+    struct raw_key_details_st raw_key;
+
+    if (skey == NULL || key == NULL || len == NULL) {
+        ERR_raise(ERR_LIB_EVP, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+
+    raw_key.key = (const void **)key;
+    raw_key.len = len;
+
+    return evp_skeymgmt_export(skey->skeymgmt, skey->keydata,
+                               OSSL_SKEYMGMT_SELECT_SECRET_KEY,
+                               get_secret_key, &raw_key);
+}
+
+EVP_SKEY *EVP_SKEY_import_raw_key(OSSL_LIB_CTX *libctx, const char *skeymgmtname,
+                                  unsigned char *key, size_t keylen,
+                                  const char *propquery)
+{
+    OSSL_PARAM params[2];
+
+    params[0] = OSSL_PARAM_construct_octet_string(OSSL_SKEY_PARAM_RAW_BYTES,
+                                                  (void *)key, keylen);
+    params[1] = OSSL_PARAM_construct_end();
+
+    return EVP_SKEY_import(libctx, skeymgmtname, propquery,
+                           OSSL_SKEYMGMT_SELECT_SECRET_KEY, params);
+}
+
+int EVP_SKEY_up_ref(EVP_SKEY *skey)
+{
+    int i;
+
+    if (CRYPTO_UP_REF(&skey->references, &i) <= 0)
+        return 0;
+
+    REF_PRINT_COUNT("EVP_SKEY", i, skey);
+    REF_ASSERT_ISNT(i < 2);
+    return ((i > 1) ? 1 : 0);
```

Comment:
Please drop the extra parentheses (there should be none).

Reply:
Done, thanks